### PR TITLE
Kerr limit added

### LINKED ISSUE #109: Coding the right event horizon for Kerr & Kerr-Newman space-time
🐞 **Problem**
As of now, in calculation of time-like trajectory in Kerr metric stops and raises a warning on reaching schwarzschild radius. But event horizon in kerr spacetime is smaller than schwarzschild radius. But the exact values need to be researched.

🎯 **Goal**
  - Research about the problem and specify the correct radius after which computation should stop in `metric/kerr.py`.

- [x] Kerr

- [ ] Kerr-Newman

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #109`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

### PR DESCRIPTION
**Pull request by shilpi1958**
Fixes #109 

@ritzvik @shreyasbapat

### CODE REVIEW

File: src/einsteinpy/metric/kerr.py [hunk 1]
```diff
@@ -32,6 +32,7 @@ def __init__(self, pos_vec, vel_vec, time, M, a):
             (self.time.value, self.pos_vec, self.time_vel.value, self.vel_vec)
         )
         self.schwarzschild_r = scr(M)
+        self.kerr_r = kerr_utils.event_horizon(self.schwarzschild_r, self.a, theta=np.pi / 2)
```
Comment (ritzvik) [inline src/einsteinpy/metric/kerr.py]: There is no need for the class variable `self.kerr_r` I believe

File: src/einsteinpy/metric/kerr.py [hunk 2]
```diff
@@ -123,32 +124,32 @@ def f_vec(self, ld, vec):
         for i in range(4):
             vals[i] = vec[i + 4]
         vals[4] = -2.0 * (
-            chl[0][0][1] * vec[4] * vec[5]
-            + chl[0][0][2] * vec[4] * vec[6]
-            + chl[0][1][3] * vec[5] * vec[7]
-            + chl[0][2][3] * vec[6] * vec[7]
+            chl[0,0,1] * vec[4] * vec[5]
```
Comment (ritzvik) [inline src/einsteinpy/metric/kerr.py]: Great ! This went unnoticed

File: src/einsteinpy/metric/kerr.py [hunk 3]
```diff
@@ -193,14 +194,15 @@ def calculate_trajectory(
             t_bound=end_lambda,
             **OdeMethodKwargs
         )
-        _scr = self.schwarzschild_r.value * 1.001
+      # _scr = self.schwarzschild_r.value * 1.001
+        _kerr_r = self.kerr_r.value * 1.001
```
Comment (ritzvik) [inline src/einsteinpy/metric/kerr.py]: Can you name call it like `_event_hor = kerr_utils.event_horizon(self.schwarzschild_r.value, self.a) * 1.001

File: src/einsteinpy/metric/kerr.py [hunk 4]
```diff
@@ -265,6 +267,7 @@ def calculate_trajectory_iterator(
             **OdeMethodKwargs
         )
         _scr = self.schwarzschild_r.value * 1.001
+        _kerr_r = kerr*1.001
```
Comment (ritzvik) [inline src/einsteinpy/metric/kerr.py]: Here also !

File: src/einsteinpy/metric/kerr.py [hunk 5]
```diff
@@ -277,9 +280,9 @@ def yielder_func():
                     temp[5:8] = BLToCartesian_vel(ODE.y[1:4], ODE.y[5:8], self.a)
                     yield (ODE.t, temp)
                 ODE.step()
-                if (not singularity_reached) and (ODE.y[1] <= _scr):
+                if (not singularity_reached) and (ODE.y[1] <= _kerr_r):
```
Comment (ritzvik) [inline src/einsteinpy/metric/kerr.py]: Change the variables accordingly 🙂

**Comment by ritzvik** [comment]
@shilpi1958 
Great efforts. I have requested some changes. Do see to them. 
We also use `black` and `isort` to fomat the codes.
After you have made the changes, do use them.
Also, I believe that some tests also need changes. I would be handling the tests.

Finally, thanks @shilpi1958

File: src/einsteinpy/metric/kerr.py (hunk 3 above)
Comment (shilpi1958) [reply src/einsteinpy/metric/kerr.py]: yeah I should have done this!

**Comment by shreyasbapat** [comment]
Thanks for the Pull Request @shilpi1958  :)
I will review the code thoroughly once the quality checks pass on your code! 
https://circleci.com/gh/einsteinpy/einsteinpy/1637?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by shreyasbapat** [comment]
Also, I have taken a little liberty to edit the PR Description to fix the hyperlink to the issue with your permission :)

**Comment by ritzvik** [comment]
@shilpi1958 could you run `black` and `isort` on this. I believe you would already have them if you have the `dev` version of einsteinpy installed. Otherwise, they are available on pip.

**Comment by shilpi1958** [comment]
Okay, I will try running black and isort and update this!

**Comment by ritzvik** [comment]
Changes required in `tests/test_metric/test_kerr.py` :
  - Replace the values in lines `23, 37 and 126` with `2e-3`. The current values do not take into consideration the maximum spin factor possible.

**Comment by shilpi1958** [comment]
made the changes in test_kerr.py

**Comment by ritzvik** [comment]
> made the changes in test_kerr.py

I suspect einsteinpy is not properly set up on your system. Therefore, `isort` isn't working. See https://docs.einsteinpy.org/en/latest/getting_started.html#development-version

For the time being, just insert a newline after `import numpy as np`, it should pass the tests.

File: src/einsteinpy/metric/kerr.py [hunk 6]
```diff
@@ -194,13 +193,16 @@ def calculate_trajectory(
...
+        )
         while ODE.t < end_lambda:
             vec_list.append(ODE.y)
             lambda_list.append(ODE.t)
             ODE.step()
-            if (not singularity_reached) and (ODE.y[1] <= _scr):
+            if (not singularity_reached) and (ODE.y[1] <= _event_hor):
                 warnings.warn(
-                    "r component of position vector reached Schwarzchild Radius. ",
+                    "r component of position vector reached Kerr Radius. ",
```
Comment (ritzvik) [inline src/einsteinpy/metric/kerr.py]: One more change,
Could you change this to `r component of position vector reached event horizon.  `

File: src/einsteinpy/metric/kerr.py [hunk 7]
```diff
@@ -277,9 +282,9 @@ def yielder_func():
                     temp[5:8] = BLToCartesian_vel(ODE.y[1:4], ODE.y[5:8], self.a)
                     yield (ODE.t, temp)
                 ODE.step()
-                if (not singularity_reached) and (ODE.y[1] <= _scr):
+                if (not singularity_reached) and (ODE.y[1] <= _event_hor):
                     warnings.warn(
-                        "r component of position vector reached Schwarzchild Radius. ",
+                        "r component of position vector reached Kerr Radius. ",
```
Comment (ritzvik) [inline src/einsteinpy/metric/kerr.py]: Here also.

**Comment by shilpi1958** [comment]
I install dev environment and then ran isort. 
sorry for all the trouble :(

**Comment by shreyasbapat** [comment]
> I just installed dev environment and then ran isort.
> sorry for all the trouble :(

It's fine ;)

**Comment by shilpi1958** [comment]
Yay!

File: src/einsteinpy/tests/test_metric/test_kerr.py [hunk 8]
```diff
@@ -123,7 +123,7 @@ def test_calculate_trajectory3():
             [0 * u.m / u.s, 0.1 * u.rad / u.s, 951.0 * u.rad / u.s],
             0 * u.s,
             4e24 * u.kg,
-            0.33,
+            2e-3,
```
Comment (shreyasbapat) [inline src/einsteinpy/tests/test_metric/test_kerr.py]: Why did you change these?

Comment (shilpi1958) [reply src/einsteinpy/tests/test_metric/test_kerr.py]: @ritzvik  made a comment about this above.
I shouldn't have changed these?

Comment (ritzvik) [reply src/einsteinpy/tests/test_metric/test_kerr.py]: It was requested by me. Earlier the values were not taking into consideration the max spin factor.

Comment (shreyasbapat) [reply src/einsteinpy/tests/test_metric/test_kerr.py]: Seems like @ritzvik Asked you to change it! :) Cool

Comment (ritzvik) [reply src/einsteinpy/tests/test_metric/test_kerr.py]: > @ritzvik made a comment about this above.
> I shouldn't have changed these?

Conveyed this to @shreyasbapat . No problem.
